### PR TITLE
gh-126156: Improve performance of creating `Morsel` objects

### PR DESCRIPTION
**Pull request by bdraco**
Replaces the manually constructed loop with using the `dict.update` on a pre-constructed `dict`

This works out to 3.8x faster, benchmark below

* Issue: gh-126156

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-10-30-00-12-22.gh-issue-126156.BOSqv0.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Improved performance of creating :py:class:`~http.cookies.Morsel` objects
```
Comment (bdraco) [inline Misc/NEWS.d/next/Library/2024-10-30-00-12-22.gh-issue-126156.BOSqv0.rst]:
```suggestion
Improved performance of creating :py:class:`~http.cookies.Morsel` objects.
```

File: Lib/http/cookies.py [hunk 2]
```diff
@@ -266,15 +266,16 @@ class Morsel(dict):
         "samesite" : "SameSite",
     }
 
+    _reserved_defaults = {key: "" for key in _reserved}
```
Comment (picnixz) [inline Lib/http/cookies.py]: You can use `dict.fromkeys(_reserved, "")`

**Review by picnixz — COMMENTED** [review]
Can we have benchmarks using pyperf instead of timeit please? In addition can you tell us whether benchmarks are using a debug build, a release build or a PGO build? (the last one would be the realistic one). Thanks!

**Comment by ZeroIntensity** [comment]
I doubt we'll see much of a speedup on pyperformance, this is a tiny change.

**Comment by picnixz** [comment]
> I doubt we'll see much of a speedup on pyperformance

I wasn't talking about pyperformance but with pyperf. I want to know whether it's a release or debug build as well (this matters a lot). In addition, this could help us in quantizing the precise improvements in a NEWS entry (using the geometric mean, not just max-comparisons).

File: Lib/http/cookies.py (hunk 2 above)
Comment (bdraco) [inline Lib/http/cookies.py]:
```suggestion
    _reserved_defaults = dict.fromkeys(_reserved, "")
```

**Review by picnixz — COMMENTED** [review]
I assume that those numbers are for a release + PGO build and they look great!  Overall, using a class variable for defaults reduces the variation in the running time so it's better as well IMO. The numbers I got are as follows:

### Using `dict.update(self, self._reserved_defaults)`


```text
.....................
Morsel (ref): Mean +- std dev: 484 ns +- 11 ns
.....................
Morsel (new): Mean +- std dev: 124 ns +- 3 ns
```


```text
+-----------+--------+----------------------+
| Benchmark | ref    | new                  |
+===========+========+======================+
| Morsel    | 484 ns | 124 ns: 3.89x faster |
+-----------+--------+----------------------+
```

### Using `self |= self._reserved_defaults`

```text
.....................
Morsel (ref): Mean +- std dev: 477 ns +- 13 ns
.....................
Morsel (new): Mean +- std dev: 106 ns +- 2 ns
```

```
+-----------+--------+----------------------+
| Benchmark | ref    | new                  |
+===========+========+======================+
| Morsel    | 477 ns | 106 ns: 4.52x faster |
+-----------+--------+----------------------+
```

So we can be a bit more faster if we use the `|=` operator instead (I think we can also update the NEWS entry to include the improvement factor which is 4.5 in non-debug optimized builds).

File: Lib/http/cookies.py [hunk 3]
```diff
@@ -266,15 +266,16 @@ class Morsel(dict):
...
     _flags = {'secure', 'httponly'}
 
     def __init__(self):
         # Set defaults
         self._key = self._value = self._coded_value = None
 
         # Set default attributes
-        for key in self._reserved:
-            dict.__setitem__(self, key, "")
+        dict.update(self, self._reserved_defaults)
```
Comment (bdraco) [inline Lib/http/cookies.py:278]:
```suggestion
        self |= self._reserved_defaults
```

File: Misc/NEWS.d/next/Library/2024-10-30-00-12-22.gh-issue-126156.BOSqv0.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Improved performance of creating :py:class:`~http.cookies.Morsel` objects.
```
Comment (bdraco) [inline Misc/NEWS.d/next/Library/2024-10-30-00-12-22.gh-issue-126156.BOSqv0.rst]:
```suggestion
Improved performance of creating :py:class:`~http.cookies.Morsel` objects by a factor of 4.5x.
```

**Comment by bdraco** [comment]
> I assume that those numbers are for a release + PGO build and they look great!

Yes, they are

> So we can be a bit more faster if we use the `|=` operator instead (I think we can also update the NEWS entry to include the improvement factor which is 4.5 in non-debug optimized builds).

I made the adjustment and updated the NEWS entry

**Comment by picnixz** [comment]
Ah! I just wondered something. Is there *any* reason to use `dict.update` in order to by-pass a custom `__ior__` actually? because that's what the current implementation does (namely bypassing a custom `update`).

**Review by ZeroIntensity — APPROVED** [review]
LGTM. Typically, we don't like to accept changes for performance increases on pure-Python code, but I think this one is small enough that it's ok.

**Comment by bdraco** [comment]
> Ah! I just wondered something. Is there _any_ reason to use `dict.update` in order to by-pass a custom `__ior__/update` actually? because that's what the current implementation does (namely bypassing a custom `update`).

AFAICT it would only be a concern for `.update()` since custom `__ior__` isn't implemented in `Morsel`

**Comment by picnixz** [comment]
> LGTM. Typically, we don't like to accept changes for performance increases on pure-Python code

Isn't it assuming that there is no C implementation (and I don't think there is, or is there?)?

> AFAICT it would only be a concern for .update() since custom __ior__ isn't implemented in Morsel

Yes, but what about a user overriding it? maybe we should keep `dict.update` in the constructor?

File: Misc/NEWS.d/next/Library/2024-10-30-00-12-22.gh-issue-126156.BOSqv0.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Improved performance of creating :py:class:`~http.cookies.Morsel` objects by a factor of 4.5x.
```
Comment (bdraco) [inline Misc/NEWS.d/next/Library/2024-10-30-00-12-22.gh-issue-126156.BOSqv0.rst]:
```suggestion
Improved performance of creating :py:class:`~http.cookies.Morsel` objects by a factor of 3.8x.
```

**Comment by bdraco** [comment]
I guess someone might subclass it so I reverted to the `dict.update` call

File: Misc/NEWS.d/next/Library/2024-10-30-00-12-22.gh-issue-126156.BOSqv0.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Improved performance of creating :py:class:`~http.cookies.Morsel` objects by a factor of 3.8x.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-10-30-00-12-22.gh-issue-126156.BOSqv0.rst]:
```suggestion
Improve performances of creating :py:class:`~http.cookies.Morsel` objects by a factor of 3.8x.
```

**Review by picnixz — APPROVED** [review]
Thanks for that and sorry for the back and forth!

**Comment by bdraco** [comment]
> Thanks for that and sorry for the back and forth!

No worries. It was quite constructive. Thanks.

**Comment by hauntsaninja** [comment]
> we don't like to accept changes for performance increases on pure-Python code

@ZeroIntensity we do accept performance improvements in pure Python code. They just a) need to either be substantial or in very hot code, b) not hurt readability, c) not meaningfully change semantics. We also consider ourselves to be not perfect judges of "not meaningfully change semantics" (this is e.g. an important reason why we reject cosmetic changes / refactors)

This case is motivated by real-world profiling in addition to having a meaningful result on microbenchmark, and is arguably more readable. All to say, thanks for a nice change! :-)

**Review by hauntsaninja — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks

**Pull request merged**